hips/hip-0019.md

```diff
@@ -0,0 +1,186 @@
+---
+hip: 9999
+title: "Add support for an images-lock.yml file"
+authors: [ "Javier Salmerón García jsalmeron@vmware.com", "Martín Pérez martinpe@vmware.com", "Pablo Caraballo Llorente pcaraballo@vmware.com" ]
+created: "2022-12-16"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+This HIP is built on top [HIP-0015](https://github.com/helm/community/blob/main/hips/hip-0015.md), taking it a step further.
+
+While the original document described a standard way of declaring Helm Charts' images through `annotations`, here we propose a complementing way of generating a complete snapshot view of all the container images existing at the time of packaging a Helm Chart: **an images lock file**. This file would be uploaded to OCI registries alongside the Charts, providing both evidence and inventory data needed to simplify the process of moving Helm Charts between registries and to simplify Helm Chart deployments on air gapped environments.
+
+## Motivation
+
+Container images are stored separately from Helm Charts. On one hand, this makes the Helm Chart really small in size. But on the other hand, this brings many challenges in terms of distribution and network access to images, which could be avoided if the application was self-contained.
```

Comment:
From my experience, there are two differing ways helm charts are used.

1. The chart "floats" around the application. In that helm chart version X, can be used to install a range of application versions Y... e.g. https://artifacthub.io/packages/helm/kubegemsapp/nginx#nginx-parameters has the `image.tag` property. Which effectively corresponds to the version of nginx being deployed (and which has little correspondence to the helm chart version there)

2. The chart version is synonymous with the application version. ie. helm chart version X is used to deploy application version X. For which the chart must not allow parameterization of image tags. I don't know of an example in the wild, but we follow this pattern internally at my company for delivering software.

I think for this HIP to practically succeed, there needs to be a way to account for the first scenario. But don't get me wrong, also enhancing/formalizing a methology for the second scenario would be fantastic. Very glad to see this HIP!

Reply:
This is great feedback @gjenkins8. I don't see much impediment in this proposal for 1. or 2. Basically what we are asking for Helm ecosystem is to have a way to generate a bill of materials at some agreed lifecycle command (e.g. helm dependency update). That command for example generates today a Chart.lock but in the future it could further decorate that lock or generate a different listing the container image dependencies. 

Those container image annotations don't need to point to an specific version. e.g. a chart using nginx container could declare de dependency to `1.23.3`, `1.23`, `1`, or not declare it at all and use `latest`. The lock file just states what was available at the given tag in dependency update time. As a Helm Chart is just a recipe, an user should always be able to use `--set tag.image=1.12` and get an nginx deployment with nginx 1.12. 

Now, lets say some new OCI/relocation/deployment/verification tooling exists built on top of this HIP and an user wants to use the latest nginx Helm Chart with nginx 1.12. That user, or automation, could simply change the annotation and pin that Helm Chart to that specific container image, or it could manually update the images lock (discouraged). Perhaps there are even more ways to do this.